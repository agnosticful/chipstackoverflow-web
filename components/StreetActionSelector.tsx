import { Radio } from "antd";
import * as React from "react";
import styled from "styled-components";

interface Props extends React.Attributes {
  /** The bet size that the current aggressor made. */
  tableMaxBetSize?: number;
  /** The bet size that the same player already made.
   * If it's preflop and the player seats at BB, the value should be `1`.
   * `previousBetSize` must be less than or equal `tableMaxBetSize`.
   **/
  previousBetSize?: number;
  /** The next minimum raise size will be `tableMaxBetSize` + `minBetSizeDiff`.
   * In general, `previousBetSize` is supposed to be `tableMaxBetSize` - previous `tableMaxBetSize`.
   **/
  minBetSizeDiff?: number;
  onChange?: (betSize: number) => void;
  className?: string;
  style?: React.CSSProperties;
}

export default function StreetActionSelector({
  tableMaxBetSize = 0,
  previousBetSize = 0,
  minBetSizeDiff = 1,
  onChange = () => undefined,
  ...props
}: Props) {
  if (previousBetSize > tableMaxBetSize) {
    throw new Error(
      "previousBetSize must be less than or equal tableMaxBetSize."
    );
  }

  const inputRef = React.useRef<HTMLInputElement>(null);
  const [action, setAction] = React.useState<StreetAction | null>(null);

  return (
    <Radio.Group
      value={action}
      size="large"
      onChange={e => {
        const action = e.target.value;

        setAction(action);

        let newBetSize: number;

        switch (action) {
          case StreetAction.fold:
            newBetSize = previousBetSize;
            break;
          case StreetAction.checkOrCall:
            newBetSize = tableMaxBetSize;
            break;
          case StreetAction.betOrRaise:
            newBetSize = parseFloat(inputRef.current!.value);
            break;
        }

        onChange(newBetSize!);
      }}
      {...props}
    >
      <Radio.Button value={StreetAction.fold}>Fold</Radio.Button>

      <Radio.Button value={StreetAction.checkOrCall}>
        {tableMaxBetSize === previousBetSize ? "Check" : "Call"}
      </Radio.Button>

      <BetButton
        value={StreetAction.betOrRaise}
        onClick={() => {
          inputRef.current!.focus();
        }}
        tabIndex={action === StreetAction.betOrRaise ? -1 : 0}
        active={action === StreetAction.betOrRaise}
      >
        {tableMaxBetSize === 0 ? "Bet" : "Raise"}

        <BetSizeInput
          type="number"
          min={tableMaxBetSize + minBetSizeDiff}
          max={99999999}
          defaultValue={tableMaxBetSize + minBetSizeDiff}
          placeholder={`${tableMaxBetSize}`}
          tabIndex={action === StreetAction.betOrRaise ? 0 : -1}
          onChange={e => {
            if (action !== StreetAction.betOrRaise) return;

            onChange(parseFloat(e.currentTarget.value));
          }}
          ref={inputRef}
          key={`${tableMaxBetSize}`}
          active={action === StreetAction.betOrRaise}
        />

        {action === StreetAction.betOrRaise ? "BB" : null}
      </BetButton>
    </Radio.Group>
  );
}

enum StreetAction {
  fold = "f",
  checkOrCall = "c",
  betOrRaise = "r"
}

const BetButton = styled(Radio.Button)<{ active: boolean }>``;

const BetSizeInput = styled.input<{ active: boolean }>`
  width: ${({ active }) => (active ? "96px" : "0")};
  padding: ${({ active }) => (active ? "0 4px 0 15px" : "0")};
  border: none;
  box-shadow: none;
  outline: none;
  text-align: right;
  opacity: ${({ active }) => (active ? 1 : 0)};
  transition: width 150ms ease-in-out, padding 150ms ease-in-out,
    opacity 150ms ease-in-out;
`;
