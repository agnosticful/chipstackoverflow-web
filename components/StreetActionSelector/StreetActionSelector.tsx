import * as React from "react";
import styled from "styled-components";
import { GameStreetActionType } from "../../models/GameSituation";

interface Props extends React.Attributes {
  /**
   * The bet size that the current aggressor made.
   * */
  tableMaxBetSize?: number;
  /**
   * The bet size that the same player already made.
   * If it's preflop and the player seats at BB, the value should be `1`.
   * `previousBetSize` must be less than or equal `tableMaxBetSize`.
   **/
  previousBetSize?: number;
  onChange?: (type: GameStreetActionType, betSize: number) => void;
  className?: string;
  style?: React.CSSProperties;
}

export default function GameStreetActionTypeSelector({
  tableMaxBetSize = 0,
  previousBetSize = 0,
  onChange = () => undefined,
  ...props
}: Props) {
  if (previousBetSize > tableMaxBetSize) {
    throw new Error(
      "previousBetSize must be less than or equal tableMaxBetSize."
    );
  }

  const inputRef = React.useRef<HTMLInputElement>(null);
  const [action, setAction] = React.useState<GameStreetActionType | null>(null);

  const onFoldButtonClick = React.useCallback(() => {
    setAction(GameStreetActionType.fold);
    onChange(GameStreetActionType.fold, previousBetSize);
  }, [previousBetSize]);

  const onCheckOrCallButtonClick = React.useCallback(() => {
    const action =
      tableMaxBetSize === 0
        ? GameStreetActionType.check
        : GameStreetActionType.call;

    setAction(action);
    onChange(action, tableMaxBetSize);
  }, [tableMaxBetSize]);

  const onBetOrRaiseButtonClick = React.useCallback(() => {
    const action =
      tableMaxBetSize === 0
        ? GameStreetActionType.bet
        : GameStreetActionType.raise;

    setAction(action);
    onChange(action, parseFloat(inputRef.current!.value));

    inputRef.current!.focus();
  }, []);

  const onBetSizeChange = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) =>
      onChange(
        tableMaxBetSize === 0
          ? GameStreetActionType.bet
          : GameStreetActionType.raise,
        parseFloat(e.currentTarget.value)
      ),
    []
  );

  return (
    <Root {...props}>
      <SelectableButton
        onClick={onFoldButtonClick}
        active={action === GameStreetActionType.fold}
        data-test-id="fold-button"
      >
        Fold
      </SelectableButton>

      <SelectableButton
        onClick={onCheckOrCallButtonClick}
        active={
          action === GameStreetActionType.check ||
          action === GameStreetActionType.call
        }
        data-test-id="check-or-call-button"
      >
        {tableMaxBetSize === previousBetSize ? "Check" : "Call"}
      </SelectableButton>

      <SelectableButton
        tabIndex={
          action === GameStreetActionType.bet ||
          action === GameStreetActionType.raise
            ? -1
            : 0
        }
        onClick={onBetOrRaiseButtonClick}
        active={
          action === GameStreetActionType.bet ||
          action === GameStreetActionType.raise
        }
        data-test-id="bet-or-raise-button"
      >
        {tableMaxBetSize === 0 ? "Bet" : "Raise"}

        <BetSizeInput
          type="number"
          min={0}
          max={99999999}
          defaultValue={tableMaxBetSize}
          placeholder={`${tableMaxBetSize}`}
          tabIndex={
            action === GameStreetActionType.bet ||
            action === GameStreetActionType.raise
              ? 0
              : -1
          }
          onChange={onBetSizeChange}
          ref={inputRef}
          key={`${tableMaxBetSize}`}
          active={
            action === GameStreetActionType.bet ||
            action === GameStreetActionType.raise
          }
          data-test-id="bet-size-input"
        />

        {action === GameStreetActionType.bet ||
        action === GameStreetActionType.raise
          ? "BB"
          : null}
      </SelectableButton>
    </Root>
  );
}

const Root = styled.div`
  display: inline-block;
  border-radius: 4px;
  box-shadow: 0px 0px 8px #222f3e1f, 0px 8px 16px #222f3e0f;
  transition: box-shadow 200ms ease, transform 200ms ease;

  :hover {
    box-shadow: 0px 0px 8px #222f3e3f, 0px 8px 16px #222f3e1f;
    transform: translate3d(0px, -2px, 0px);
  }
`;

const SelectableButton = styled.button<{ active: boolean }>`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  min-width: 96px;
  height: 32px;
  padding: 0 12px;
  background: #0f151c;
  border-color: transparent;
  border-style: solid;
  border-top-width: 1px;
  border-bottom-width: 1px;
  border-left-width: 1px;
  border-right-width: 0px;
  outline: none;
  color: #fff;
  font-size: 14px;
  -webkit-font-smoothing: antialiased;
  cursor: pointer;
  transition: width 200ms ease, background 200ms ease, border 200ms ease,
    color 200ms ease;

  :first-child {
    border-top-left-radius: 4px;
    border-bottom-left-radius: 4px;
  }

  :last-child {
    border-right-width: 1px;
    border-top-right-radius: 4px;
    border-bottom-right-radius: 4px;
  }

  :hover,
  :focus {
    background: #fff;
    border-color: #0f151c;
    color: #0f151c;
  }

  ${({ active }) =>
    active
      ? `background: #fff;
         border-color: #0f151c;
         color: #0f151c;`
      : ""}
`;

const BetSizeInput = styled.input<{ active: boolean }>`
  box-sizing: border-box;
  width: ${({ active }) => (active ? "80px" : "0")};
  background: transparent;
  line-height: 30px;
  padding: ${({ active }) => (active ? "0 8px" : "0")};
  border: none;
  outline: none;
  color: #0f151c;
  font-size: 14px;
  text-align: right;
  transition: width 200ms ease, padding 200ms ease;

  ::placeholder {
    color: ${({ active }) => (active ? "#c8d6e5" : "transparent")};
    transition: color 200ms ease;
  }

  ::-webkit-outer-spin-button,
  ::-webkit-inner-spin-button {
    appearance: none;
    margin: 0;
  }

  &[type="number"] {
    -moz-appearance: textfield;
  }
`;
