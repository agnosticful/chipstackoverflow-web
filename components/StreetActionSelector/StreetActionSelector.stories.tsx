import { action } from "@storybook/addon-actions";
import { number } from "@storybook/addon-knobs";
import * as React from "react";
import StreetActionSelector from "./StreetActionSelector";

export default {
  title: "StreetActionSelector",
  component: StreetActionSelector
};

export const example = () => {
  const tableMaxBetSize = number("tableMaxBetSize", undefined as any, {
    min: 0
  });

  return (
    <StreetActionSelector
      tableMaxBetSize={tableMaxBetSize}
      previousBetSize={number("previousBetSize", undefined as any, {
        min: 0,
        max: tableMaxBetSize
      })}
      onChange={action("onChange")}
    />
  );
};

export const whenSomeoneAlreadyBet = () => (
  <StreetActionSelector tableMaxBetSize={22.5} onChange={action("onChange")} />
);

export const atBigBlind = () => (
  <StreetActionSelector
    tableMaxBetSize={1}
    previousBetSize={1}
    onChange={action("onChange")}
  />
);
