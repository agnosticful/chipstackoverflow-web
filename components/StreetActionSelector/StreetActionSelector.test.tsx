import * as React from "react";
import { act, create } from "react-test-renderer";
import { createRenderer } from "react-test-renderer/shallow";
import StreetActionSelector from "./StreetActionSelector";

describe("<StreetActionSelector>", () => {
  it("matches with the previous snapshot", () => {
    const renderer = createRenderer();

    renderer.render(<StreetActionSelector />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("matches with the previous snapshot (someone raised)", () => {
    const onChange = jest.fn().mockName("onChange");
    const renderer = createRenderer();

    renderer.render(
      <StreetActionSelector tableMaxBetSize={22.5} onChange={onChange} />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("matches with the previous snapshot (big blind in preflop)", () => {
    const onChange = jest.fn().mockName("onChange");
    const renderer = createRenderer();

    renderer.render(
      <StreetActionSelector
        tableMaxBetSize={1}
        previousBetSize={1}
        onChange={onChange}
      />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("throws when props.previousBetSize is more than props.tableMaxBetSize", async () => {
    const onChange = jest.fn().mockName("onChange");
    const renderer = createRenderer();

    expect(() =>
      renderer.render(
        <StreetActionSelector
          tableMaxBetSize={1}
          previousBetSize={2}
          onChange={onChange}
        />
      )
    ).toThrow();
  });

  it('calls onChange() with props.previousBetSize when "Fold" button is clicked', async () => {
    const onChange = jest.fn().mockName("onChange");
    const renderer = create(
      <StreetActionSelector
        tableMaxBetSize={1}
        previousBetSize={0}
        onChange={onChange}
      />
    );

    const foldButton = renderer.root.find(
      node => node.props["data-test-id"] === "fold-button"
    );

    await act(async () => {
      foldButton.props.onClick();
    });

    await act(async () => {
      renderer.update(
        <StreetActionSelector
          tableMaxBetSize={3}
          previousBetSize={1}
          onChange={onChange}
        />
      );
    });

    await act(async () => {
      foldButton.props.onClick();
    });

    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenNthCalledWith(1, 0);
    expect(onChange).toHaveBeenNthCalledWith(2, 1);
  });

  it('calls onChange() with props.tableMaxBetSize when "Call" button is clicked', async () => {
    const onChange = jest.fn().mockName("onChange");
    const renderer = create(
      <StreetActionSelector
        tableMaxBetSize={1}
        previousBetSize={0}
        onChange={onChange}
      />
    );

    const checkOrCallButton = renderer.root.find(
      node => node.props["data-test-id"] === "check-or-call-button"
    );

    await act(async () => {
      checkOrCallButton.props.onClick();
    });

    await act(async () => {
      renderer.update(
        <StreetActionSelector
          tableMaxBetSize={3}
          previousBetSize={1}
          onChange={onChange}
        />
      );
    });

    await act(async () => {
      checkOrCallButton.props.onClick();
    });

    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenNthCalledWith(1, 1);
    expect(onChange).toHaveBeenNthCalledWith(2, 3);
  });

  it("sets defaultValue at bet size input that is sum of props.tableMaxBetSize", async () => {
    const renderer = create(
      <StreetActionSelector tableMaxBetSize={1} previousBetSize={0} />
    );

    let betSizeInput = renderer.root.find(
      node => node.props["data-test-id"] === "bet-size-input"
    );

    expect(betSizeInput.props.defaultValue).toBe(1);

    await act(async () => {
      renderer.update(
        <StreetActionSelector tableMaxBetSize={3} previousBetSize={1} />
      );
    });

    betSizeInput = renderer.root.find(
      node => node.props["data-test-id"] === "bet-size-input"
    );

    expect(betSizeInput.props.defaultValue).toBe(3);
  });

  it('calls onChange() with the bet size input\'s value when "Raise" button is clicked', async () => {
    const onChange = jest.fn().mockName("onChange");
    const betSizeInputValue = 123456789;
    const renderer = create(
      <StreetActionSelector
        tableMaxBetSize={1}
        previousBetSize={0}
        onChange={onChange}
      />,
      {
        createNodeMock: element => {
          if (element.props["data-test-id"] === "bet-size-input") {
            return {
              focus: () => undefined,
              value: betSizeInputValue
            };
          }

          return null;
        }
      }
    );

    const betOrRaiseButton = renderer.root.find(
      node => node.props["data-test-id"] === "bet-or-raise-button"
    );

    await act(async () => {
      betOrRaiseButton.props.onClick();
    });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenNthCalledWith(1, betSizeInputValue);
  });

  it('sets a focus on bet size input when "Raise" button is clicked', async () => {
    const focus = jest.fn().mockName("focus");
    const renderer = create(
      <StreetActionSelector tableMaxBetSize={1} previousBetSize={0} />,
      {
        createNodeMock: element => {
          if (element.props["data-test-id"] === "bet-size-input") {
            return {
              focus,
              value: null
            };
          }

          return null;
        }
      }
    );

    const betOrRaiseButton = renderer.root.find(
      node => node.props["data-test-id"] === "bet-or-raise-button"
    );

    await act(async () => {
      betOrRaiseButton.props.onClick();
    });

    expect(focus).toHaveBeenCalledTimes(1);
  });

  it("calls onChange() with the value of input element whenever its value changes", async () => {
    const onChange = jest.fn().mockName("onChange");
    const renderer = create(
      <StreetActionSelector
        tableMaxBetSize={1}
        previousBetSize={0}
        onChange={onChange}
      />
    );

    let betSizeInput = renderer!.root.find(
      node => node.props["data-test-id"] === "bet-size-input"
    );

    await act(async () => {
      betSizeInput.props.onChange({ currentTarget: { value: 20 } });
    });

    await act(async () => {
      renderer.update(
        <StreetActionSelector
          tableMaxBetSize={3}
          previousBetSize={1}
          onChange={onChange}
        />
      );
    });

    // re-obtain betSizeInput element
    // because it would be remounted whenever props.tableMaxBetSize changes
    betSizeInput = renderer!.root.find(
      node => node.props["data-test-id"] === "bet-size-input"
    );

    await act(async () => {
      betSizeInput.props.onChange({ currentTarget: { value: 43 } });
    });

    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenNthCalledWith(1, 20);
    expect(onChange).toHaveBeenNthCalledWith(2, 43);
  });
});
