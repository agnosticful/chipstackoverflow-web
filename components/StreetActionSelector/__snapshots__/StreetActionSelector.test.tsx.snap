// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StreetActionSelector> matches with the previous snapshot (big blind in preflop) 1`] = `
<styled.div>
  <styled.button
    active={false}
    data-test-id="fold-button"
    onClick={[Function]}
  >
    Fold
  </styled.button>
  <styled.button
    active={false}
    data-test-id="check-or-call-button"
    onClick={[Function]}
  >
    Check
  </styled.button>
  <styled.button
    active={false}
    data-test-id="bet-or-raise-button"
    onClick={[Function]}
    tabIndex={0}
  >
    Raise
    <styled.input
      active={false}
      data-test-id="bet-size-input"
      defaultValue={1}
      max={99999999}
      min={0}
      onChange={[Function]}
      placeholder="1"
      tabIndex={-1}
      type="number"
    />
  </styled.button>
</styled.div>
`;

exports[`<StreetActionSelector> matches with the previous snapshot (someone raised) 1`] = `
<styled.div>
  <styled.button
    active={false}
    data-test-id="fold-button"
    onClick={[Function]}
  >
    Fold
  </styled.button>
  <styled.button
    active={false}
    data-test-id="check-or-call-button"
    onClick={[Function]}
  >
    Call
  </styled.button>
  <styled.button
    active={false}
    data-test-id="bet-or-raise-button"
    onClick={[Function]}
    tabIndex={0}
  >
    Raise
    <styled.input
      active={false}
      data-test-id="bet-size-input"
      defaultValue={22.5}
      max={99999999}
      min={0}
      onChange={[Function]}
      placeholder="22.5"
      tabIndex={-1}
      type="number"
    />
  </styled.button>
</styled.div>
`;

exports[`<StreetActionSelector> matches with the previous snapshot 1`] = `
<styled.div>
  <styled.button
    active={false}
    data-test-id="fold-button"
    onClick={[Function]}
  >
    Fold
  </styled.button>
  <styled.button
    active={false}
    data-test-id="check-or-call-button"
    onClick={[Function]}
  >
    Check
  </styled.button>
  <styled.button
    active={false}
    data-test-id="bet-or-raise-button"
    onClick={[Function]}
    tabIndex={0}
  >
    Bet
    <styled.input
      active={false}
      data-test-id="bet-size-input"
      defaultValue={0}
      max={99999999}
      min={0}
      onChange={[Function]}
      placeholder="0"
      tabIndex={-1}
      type="number"
    />
  </styled.button>
</styled.div>
`;
